File: go/bank_account.go
package account

import "sync"

type Account struct {
    mu sync.RWMutex
	balance int64    
    closed bool
}

func Open(amount int64) *Account {
    if amount < 0 {
        return nil
    }

    return &Account{
        balance: amount,
    }
}

func (a *Account) Balance() (int64, bool) {
    a.mu.RLock()
    defer a.mu.RUnlock()
    
    if a.closed {
        return 0, false
    }
	return a.balance, true
}

func (a *Account) Deposit(amount int64) (int64, bool) {
    a.mu.Lock()
    defer a.mu.Unlock()
    
    if a.closed {
        return 0, false
    }

	if a.balance + amount < 0 {
        return 0, false
    }

	a.balance += amount
    return a.balance, true
}

func (a *Account) Close() (int64, bool) {
    a.mu.Lock()
    defer a.mu.Unlock()
    
    if a.closed {
        return 0, false
    }
	a.closed = true
    balance := a.balance
    a.balance = 0
    return balance, true
}
gg

File: elixir/bank_account.ex
defmodule BankAccount do
  @moduledoc """
  A bank account that supports access from multiple processes.
  """

  @typedoc """
  An account handle.
  """
  @opaque account :: pid

  @doc """
  Open the bank account, making it available for further operations.
  """
  @spec open() :: account
  def open() do
    {_, pid } = Agent.start_link(fn -> 0 end)
    pid
  end

  @doc """
  Close the bank account, making it unavailable for further operations.
  """
  @spec close(account) :: any
  def close(account) do
    Agent.stop(account)
  end

  @doc """
  Get the account's balance.
  """
  @spec balance(account) :: integer | {:error, :account_closed}
  def balance(account) do
    if Process.alive?(account) do
      Agent.get(account, fn n -> n end)
    else
      {:error, :account_closed}
    end
  end

  @doc """
  Add the given amount to the account's balance.
  """
  @spec deposit(account, integer) :: :ok | {:error, :account_closed | :amount_must_be_positive}
  def deposit(_, amount) when amount <= 0 do
    {:error, :amount_must_be_positive}
  end
  def deposit(account, amount) do
    if Process.alive?(account) do
      Agent.update(account, fn n -> n + amount end)
    else
      {:error, :account_closed}
    end
  end

  @doc """
  Subtract the given amount from the account's balance.
  """
  @spec withdraw(account, integer) ::
          :ok | {:error, :account_closed | :amount_must_be_positive | :not_enough_balance}
  def withdraw(_, amount) when amount <= 0 do
    {:error, :amount_must_be_positive}
  end
  def withdraw(account, amount) do
    if Process.alive?(account) do
      Agent.get_and_update(account, fn n ->
        if n >= amount do
          {:ok, n - amount}
        else
          {{:error, :not_enough_balance}, n}
        end
      end)
    else
      {:error, :account_closed}
    end
  end
end

